enclave-runtime/src/lib.rs

```diff
@@ -578,6 +582,79 @@ where
 	Ok(())
 }
 
+//For now get the DOT/currency exchange rate from coingecko API
+#[no_mangle]
+pub unsafe extern "C" fn update_market_data_xt(
+	currency: *const u8,
+	currency_size: u32,
+	unchecked_extrinsic: *mut u8,
+	unchecked_extrinsic_size: u32,
+) -> sgx_status_t {
+	let extrinsic_slice =
+		slice::from_raw_parts_mut(unchecked_extrinsic, unchecked_extrinsic_size as usize);
+
+	let mut curr_slice = slice::from_raw_parts(currency, currency_size as usize);
+	let curr: String = Decode::decode(&mut curr_slice).unwrap();
+	let xts = match update_market_data_internal::<Block>(curr) {
+		Ok(xts) => xts,
+		Err(_) => return sgx_status_t::SGX_ERROR_UNEXPECTED,
+	};
+
+	//only one extrinsic to send over the node api directly. No need of ocall
+	let xt = match xts.get(0) {
+		Some(xt) => xt,
+		None => return sgx_status_t::SGX_ERROR_UNEXPECTED,
+	};
+
+	write_slice_and_whitespace_pad(extrinsic_slice, xt.encode());
+	sgx_status_t::SGX_SUCCESS
+}
+
+fn update_market_data_internal<PB>(curr: String) -> Result<Vec<OpaqueExtrinsic>>
+where
+	PB: BlockT<Hash = H256>,
+{
+	let signer = Ed25519Seal::unseal()?;
+	let light_client_lock = EnclaveLock::write_light_client_db()?;
+	let validator = LightClientSeal::<PB>::unseal()?;
+	let genesis_hash = validator.genesis_hash(validator.num_relays())?;
+	LightClientSeal::seal(validator)?;
+	std::mem::drop(light_client_lock);
+
+	let extrinsics_factory =
+		ExtrinsicsFactory::new(genesis_hash, signer, GLOBAL_NONCE_CACHE.clone());
+
+	//For now polkadot
+	let coin = "polkadot";
+
+	//Get the exchange rate
+	let url = CoinGeckoClient::base_url().unwrap();
+	let mut coingecko_client = CoinGeckoClient::new(url);
+	let rate = match coingecko_client.get_exchange_rate(coin, &curr) {
+		Ok(r) => r,
+		Err(e) => {
+			error!("[-] Failed to get the newest exchange rate from coingecko. {:?}", e);
+			return Err(Error::Other(e.into()))
+		},
+	};
+
+	let mut calls = Vec::<OpaqueCall>::new();
+	println!(
+		"update_market_data_internal values :coingecko value {}, currency {:?} and rate {:?}",
```

Comment:
A little bit nicer for uers that don't know the function name:

```suggestion
		"Retrieved the following values from coingecko: \n Exchange Rate: {} \n Currency: {:?} \n Rate {:?}",
```

Though I'm unsure if these values are human readable? Leaving that decision to you.

Reply:
This was to debug not for the user. I will remove it